compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp

```diff
@@ -769,6 +771,30 @@ LLVMRustOptimize(
           MPM.addPass(createModuleToFunctionPassAdaptor(AddressSanitizerPass(
               /*CompileKernel=*/false, SanitizerOptions->SanitizeAddressRecover,
               /*UseAfterScope=*/true)));
+#endif
+        }
+      );
+    }
+    if (SanitizerOptions->SanitizeKernelAddress) {
+      OptimizerLastEPCallbacks.push_back(
+        [SanitizerOptions](ModulePassManager &MPM, OptimizationLevel Level) {
+#if LLVM_VERSION_LT(15, 0)
+          MPM.addPass(RequireAnalysisPass<ASanGlobalsMetadataAnalysis, Module>());
+#endif
+#if LLVM_VERSION_GE(14, 0)
+          AddressSanitizerOptions opts = AddressSanitizerOptions{
+            /*CompileKernel=*/true,
+            SanitizerOptions->SanitizeKernelAddressRecover,
+            /*UseAfterScope=*/true,
+            AsanDetectStackUseAfterReturnMode::Runtime,
+          };
+          MPM.addPass(ModuleAddressSanitizerPass(opts));
```

Comment:
If both were combined, it would end up just enabling the `CompileKernel` parameter, which probably wouldn't work for people expecting regular ASan, but the frontend options forbid this regardless. I think it probably makes sense to combine them into a single `if` to reduce maintenance burden, so I'll go ahead and do that